### PR TITLE
Add ProtoBuf txt parsing to read max_batch_size

### PR DESCRIPTION
**Pull request by banasraf**
When config is read by the tritonserver, there is no way to determine if max_batch_size was set to 0 by the user or was it just not provided. 
To workaround it, I manually load the config text file (if it exists) and parse it (to a limited level) to read the max_batch_size field. 

Signed-off-by: Rafal <rbanas@nvidia.com>

### CODE REVIEW

File: src/config_tools/config_tools.h [hunk 1]
```diff
@@ -178,16 +178,18 @@ void ValidateInputs(TritonJson::Value &ins, const std::vector<IOConfig> &in_conf
...
-int ReadMaxBatchSize(TritonJson::Value &config);
+void ValidateConfig(TritonJson::Value &config, const std::vector<IOConfig> &in_configs,
+                    const std::vector<IOConfig> &out_configs);
 
 
 /**
- * @brief Validate the model max batch size and inputs and outputs configs against provided values.
+ * @brief Read max_batch_size set in a config represented as protocol buffer text format
+ * Returns a max_batch_size value or an empty optional when
+ * the is absent or cannot me parsed as int
```
Comment (stiepan) [inline src/config_tools/config_tools.h:190]:
```suggestion
 * the value is absent or cannot be parsed as int
```

**Comment by mzient** [comment]
Why not use RapidJSON or some other JSON parser?

File: src/config_tools/config_tools.cc [hunk 2]
```diff
@@ -421,4 +408,121 @@ void ValidateConfig(TritonJson::Value &config, const std::vector<IOConfig> &in_c
...
+  if (text.empty()) return;
+  size_t open_bracket = 0;
+  do {
+    if (text[0] == '\"') {
+      skip_string(text);
+    } else {
+      if (text[0] == bra) ++open_bracket;
+      else if (text[0] == ket) --open_bracket;
+      text.remove_prefix(1);
+    }
```
Comment (mzient) [inline src/config_tools/config_tools.cc:473]: This doesn't handle proper nesting - should be something like:
```suggestion
    if (text[0] == '\"') {
      skip_string(text);
    } else if (text[0] == '[') {
      skip_complex(text, '[', ']');
    } else if (text[0] == '{') {
      skip_complex(text, '{', '}');
    } else {
      text.remove_prefix(1);
    }
```

File: src/dali_model.h [hunk 3]
```diff
@@ -307,6 +313,7 @@ class DaliModel : public ::triton::backend::BackendModel {
   std::vector<IOConfig> pipeline_inputs_{};
   std::vector<IOConfig> pipeline_outputs_{};
   int pipeline_max_batch_size_ = -1;
+  int config_max_batch_size_ = -1;
```
Comment (mzient) [inline src/dali_model.h:316]: Why not `optional<int>` here, too?

**Review by mzient — APPROVED** [review]
With a nitpick: use `optional<int>` in the member, too.

**Comment by banasraf** [comment]
> Why not use RapidJSON or some other JSON parser?

This format is not JSON. It is [Protocol Buffer Text Format](https://developers.google.com/protocol-buffers/docs/text-format-spec)

File: src/config_tools/config_tools.cc [hunk 4]
```diff
@@ -421,4 +408,121 @@ void ValidateConfig(TritonJson::Value &config, const std::vector<IOConfig> &in_c
...
+  skip_ignored(text);
+  if (text.empty()) return {};
+  bool negative = false;
+  if (text[0] == '-') {
+    negative = true;
+    text.remove_prefix(1);
+    skip_ignored(text);
+  }
+  size_t end = 0;
+  while (end < text.size() && !(is_whitespace(text[end]) || is_sep(text[end]))) ++end;
```
Comment (stiepan) [inline src/config_tools/config_tools.cc:489]: Why `is_sep`?

**Review by stiepan — APPROVED** [review]

**Pull request merged**